Feature/support bsd date

wolfwood opened the pull request:
attempt to resolve #16 

@lindsaylevine would you be able to test this?

wolfwood committed:
first pass at date library

wolfwood committed:
fix bugs

wolfwood committed:
make get_dow() date argument optional

wolfwood committed:
updaed dates.sh to use lib-date.sh. produces identical output on linux

wolfwood committed:
Merge remote-tracking branch 'upstream/main' into feature/support-bsd-date

wolfwood commented:
looks like #15 also added a dependency on GNU date to paint.sh, so I'll need to tackle that too.

wolfwood committed:
reorganize funtions in lib-date and add time manipulation functions

wolfwood committed:
convert paint.sh to use lib-date.sh instead of calling date directly

wolfwood commented:
@BenBE would you be willing to look this over?

lindsaylevine commented:
yo just got a chance to test. checked out your branch and getting the same old error
<img width="758" alt="Screen Shot 2020-08-24 at 3 47 21 AM" src="https://user-images.githubusercontent.com/57350178/91017655-88cd7500-e5bc-11ea-8c1a-241430eb23b5.png">

rossabaker commented:
GitHub Actions supports MacOS and Windows runners. Might be a good way to catch issues like this.

wolfwood committed:
some feedback from running shellcheck

wolfwood committed:
don't add a plus sign to negative 0 (or positive 0, but I think thats OK)

wolfwood commented:
@lindsaylevine I had a older branch (now deleted) and I think you might have gotten that one, sorry. I borrowed a machine to do some some sanity testing (no git installed) and fixed another issue. was able to produce a dates.txt that matched the old one, so I think you'l be good to go this time.

You'd expect a more straightforward way to get code from a PR. I grabbed a zip file from here https://github.com/wolfwood/abolish-ice-graphiti/tree/feature/support-bsd-date

@rossabaker that's interesting. I know nothing about docker, can we change the system clock?

wolfwood commented:
@rossabaker also I started using https://www.shellcheck.net/ to review my code. it gives a lot of warnings, but if you are willing to add comments to the code to turn off the ones we don't care about we could use it as a pre-commit hook or something. I've never really used git hooks.

lindsaylevine commented:
<img width="879" alt="Screen Shot 2020-08-24 at 8 18 48 PM" src="https://user-images.githubusercontent.com/57350178/91108822-0dfd6c00-e647-11ea-9e65-adf4f741ec25.png">

dates.sh now hangs forever on this log and doesnt change dates.txt
paint.sh behaves as normal (goes through [old] dates.txt)


edit: checked git history of local test branch, 100% sure i have most recent

wolfwood commented:
@lindsaylevine its hanging on input. `./dates.sh < dow.txt > dates.txt`

lindsaylevine commented:
o 🤦‍♀️  duh. works now!

worth updating the README in this PR to reflect what needs to be run?

wolfwood commented:
@lindsaylevine yeah, README is definitely missing some important stuff. I think its also worth considering converting dates.sh to use default input and output file names that can be overridden on the command line, rather than using redirects.

I also think the larger flow can be automated more, including using the github API to make repos, switch default branches to trigger re-rendering of the contribution graph, etc, but that is worth having a discussion about on a dedicated issue.

in any case, very excited that OS X / BSD support is working!

rossabaker commented:
I use [pre-commit-hooks](https://github.com/pre-commit/pre-commit-hooks) at $WORK, and I'm sure that `shellcheck` could be one, but it's probably just about as easy to add it in a GitHub Actions CI workflow.

I don't know how to change the time in a Docker container other than to invoke `date` inside of it.  I try to pass a stubbed out clock as an argument for testing, but not so easy in a shell script.

wolfwood commented:
I'm gonna go ahead and merge this

BenBE commented on lib-date.sh:
```diff
@@ -0,0 +1,105 @@
+    echo "$newdate"
+}
+
+# not strictly speaking a year, but the begining of the github cotribution graph. 52 weeks + dow days ago
```
Should be contribution

wolfwood replied:
fixed, thanks.
